### ISSUE TITLE
Add third option "no-paren" to space-after-keywords

### ISSUE THREAD

**Issue by matoilic** [eslint/eslint#4585]
A common style for spaces after keywords is to only require them if the keyword is not followed by parentheses. This should be supported out of the box by eslint.

Good
```javascript
if(expression) { } else { }
```

Bad
```javascript
if (expression) { } else { }
```

Bad
```javascript
if(expression) { }else{ }
```

#4561 introduces a "no-paren" option to the space-after-keywords rule.

**Comment by eslintbot** [comment]
@matoilic Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
I'm not sure how common this is. Most style guides explicitly require a space after keywords to distinguish them from function calls.

@eslint/eslint-team thoughts on this?

**Comment by IanVS** [comment]
Yeah, I'd want to see some justification of calling this a "common" style.  I'm just a noob, but I've never heard of this.

**Comment by platinumazure** [comment]
I agree it's probably not nearly as common as the issue creator might claim, but is there any reason we shouldn't allow the extra option? (I'm certainly not suggesting that this should be a priority for the ESLint team to develop, but if a PR were to come in...)

**Comment by IanVS** [comment]
Personally, I think each option or rule we add to core has a cost in cognitive load for each new adopter of ESLint.  So, unless a use case or style reaches a certain subjective threshold of general applicability, I would argue it doesn't belong in core and should be a custom rule/plugin instead.

**Comment by platinumazure** [comment]
Solid argument, and you've convinced me.

**Comment by ilyavolodin** [comment]
Sorry, I might be a bit slow today, but I'm not sure I fully understand the request. So if the keyword is not followed by parentheses it shouldn't require whitespace after? Does that mean whitespaces shouldn't be required, or whitespace is not allowed?
In any case, I do agree with the rest of the comments, I'm not sure how popular this style is.

**Comment by lo1tuma** [comment]
I think this would be possible in combination with `space-before-block`, but currently we have an issue that those two rules overlap (see: #1338).

**Comment by matoilic** [comment]
The reasoning behind this style is the following:

It's common to not have a space after the `function` keyword if it's an unnamed one. So, if you don't enforce a space there, then why enforce one after the "if" or "switch" keywords. Most projects I've been part of in the past years followed that styling rule.

**Comment by nzakas** [comment]
@IanVS well said.

@platinumazure another consideration is agreeing to take on work that only a small number of people will benefit from. If we agree to do this, then it competes for attention with other issues that might be more critical. Since we have very few consistent contributors, we need to keep the backlog small and focused.

@matoilic typically because functions require no space prior to the open paren, so whether or not a function has a name doesn't alter that expectation.

That said, if @matoilic is willing to make the change, then I'm fine with accepting it.

**Comment by eTorAken** [comment]
:+1: with this issue. In our coding styles we enforce the use of no space after a `if`. I thought such a rule (`no-space-before-condition`) would already exist but it seems it does not.

**Comment by mysticatea** [comment]
I think this issue is addressed by #3869.

```json
{
    "keyword-spacing": {
        "before": true,
        "after": true,
        "overrides": {
            "if": {"after": false}
        }
    }
}
```

**Comment by nzakas** [comment]
Ah, right! Are you planning on addressing #3869 soon?

**Comment by mysticatea** [comment]
Yes, I will do within the next few days.

**Comment by nzakas** [comment]
Okay, let's leave this open to verify that #3869 covers this case as well.
